### PR TITLE
add smart fillto to enable logscale y-axis histogram barplot

### LINKED ISSUE #3000: Log scale broken for histograms
```julia
data = randn(10^6);

hist(data; axis=(;yscale=log10, ))
```

```
Failed to show value:

Found invalid y-limits (0.0f0, 280424.0f0) for scale log10 which is defined on the interval 0.0..Inf (open)

    error(::String)@error.jl:35
    autolimits(::Makie.Axis, ::Int64)@axis.jl:964
    yautolimits@axis.jl:977[inlined]

```

```julia
hist(data; axis=(;yscale=log10, limits=(nothing, nothing, 0.1, nothing)))
```

```
Failed to show value:

Found invalid y-limits (0.0f0, 280424.0f0) for scale log10 which is defined on the interval 0.0..Inf (open)

    error(::String)@error.jl:35
    autolimits(::Makie.Axis, ::Int64)@axis.jl:964
    yautolimits@axis.jl:977[inlined]
```

```julia
hist(data; axis=(;yscale=log10, limits=(nothing, nothing, 0.1, 10^6)))
```

the last one shows incorrect plot:
image

### PR DESCRIPTION
**Pull request by Moelf**
Fix #3000 

This implements what's suggested on Slack, namely, detect if transformation is one of the dangerous logs (example of non-dangerous log is `Makie.pseudolog10`), if so, use the minimal non zero value divided by 2 as the "bottom" to avoid illegal transformation `0 -> -Inf` when applying the log.

Some examples that used to error but now works:

```julia
using CairoMakie

a = "/tmp/jl_k9M7ELsL2h.png"

save(a, hist(randn(10^6); axis=(; yscale=log10)))

save(a, hist(randn(10^6); axis=(; yscale=log2)))

save(a, hist(filter!(x-> x<0 || x > 1.5, randn(10^6)); axis=(; yscale=log10)))
```

image

image
image

### CODE REVIEW

**Comment by Moelf** [comment]
(should maybe also add a test?)

File: ReferenceTests/src/tests/examples2d.jl [hunk 1]
```diff
@@ -909,6 +909,24 @@ end
     f
 end
 
+@reference_test "Log y-scale histogram (barplot)" begin
+    # make a gap in histogram as edge case
```
Comment (asinghvi17) [inline ReferenceTests/src/tests/examples2d.jl:913]:
```suggestion

**Comment by Moelf** [comment]
when you say example do you mean in the docs? I don't think we need that right?

**Comment by jkrumbiegel** [comment]
> example do you mean in the docs? I don't think we need that right?

yes a docs example would be good because we want to explain every single attribute visually in the long run. And the visual effect of `fillto = automatic` would have to be shown

**Comment by Moelf** [comment]
I would say drop zero is not needed because in both cases it should be invisible, that seems to be the only reasonable behavior

File: src/basic_recipes/barplot.jl [hunk 2]
```diff
@@ -2,6 +2,36 @@ function bar_label_formatter(value::Number)
...
+"""
+function bar_default_fillto(tf, ys, offset)
+    return ys, offset
+end
+
+# `fillto` is related to `y-axis` transofrmation only, thus we expect `tf::Tuple`
+function bar_default_fillto(tf::Tuple, ys, offset)
+    if tf[2] isa Union{typeof(log), typeof(log2), typeof(log10), Base.Fix1{typeof(log), <: Real}}
+        # use the minimal non-zero y divided by 2 as lower bound for log scale
+        smart_fillto = minimum(y -> y<=0 ? oftype(y, Inf) : y, ys) / 2
```
[the remark this answers is not in the document]
Comment (Moelf) [reply src/basic_recipes/barplot.jl:28]: This is approximately the logic of ROOT, I think visually this is appealing enough. Btw, /10 is minor ticks in log10, too many imo

**Comment by Moelf** [comment]
See the third example

**Comment by Moelf** [comment]
the last commit makes `x-scale` work too:
```julia
barplot(
           tbl.x, tbl.height,
           dodge = tbl.grp,
           color = tbl.grp,
           bar_labels = :y,
           axis = (xticks = (1:3, ["left", "middle", "right"]),
                   title = "Dodged bars horizontal with labels", xscale=log10),
           colormap = [:red, :green, :blue],
           color_over_background=:red,
           color_over_bar=:white,
           flip_labels_at=0.85,
           direction=:x,
       );
```
this used to error, but now:
image

**Comment by Moelf** [comment]
is this ready? I don't know what example I should add, `fillto` itself is documented in the `barplot` page pretty well

**Comment by Moelf** [comment]
this seems ready :eyes:

**Comment by SimonDanisch** [comment]
@jkrumbiegel wanted to do a final review before merging

**Comment by Moelf** [comment]
franklin error seems unrelated

**Comment by SimonDanisch** [comment]
Hm, a new Franklin version was released just now: https://github.com/tlienart/Franklin.jl/releases/tag/v0.10.86
I guess that's most likely the cause?

**Comment by Moelf** [comment]
yeah I think so, this PR didn't touch the the slide grid markdown

**Comment by Moelf** [comment]
Passed?

**Review by jkrumbiegel — APPROVED** [review]
Made some small changes to the test to exercise the x direction code, too. I think this is good to go now. I haven't tested out all combinations of switching back and forth in GLMakie, but it looked like we easily get limit errors that way which are not really related to the recipe, rather to the limit update code and when what value is updated. So not relevant here.

There should be a docs example added at some point, but before I spend time doing that in the current format, I'd rather wait until we switch the plotting functions to direct attribute docs like we have for Blocks. Shouldn't block this right now.
